# Application of Model Evaluation - use adult_ch6_training and adult_ch6_test data sets

In [7]:
# Importing Libraries
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.tools.tools as stattools
from sklearn.tree import DecisionTreeClassifier, export_graphviz
from sklearn.tree import plot_tree
from sklearn.metrics import confusion_matrix
import matplotlib.pyplot as plt

In [8]:
# Importing Datasets
adult_train = pd.read_csv("/Users/datascience/Desktop/ADS 502 Data Sets/Website Data Sets/adult_ch6_training.csv")
adult_test = pd.read_csv("/Users/datascience/Desktop/ADS 502 Data Sets/Website Data Sets/adult_ch6_test.csv")

## 23. Using the training dataset, create a C5.0 model (Model1) to predict a customer’s Income using Marital Status and Capital Gains and Losses. Obtain the predicted responses.

In [9]:
y = adult_train[['Income']]
X = adult_train[['Marital status', 'Cap_Gains_Losses']]

# Convert categorical variable into a dummy variable form
mar_dummy = pd.get_dummies(X['Marital status'])

#Add dummy variables back into X variables
X = pd.concat((X[['Cap_Gains_Losses']], mar_dummy), axis = 1) #Column-wise Concatation (axis=1)

# Use C5.0 model on training set (criterion = 'entropy')
c5 = DecisionTreeClassifier(criterion='entropy', max_leaf_nodes=5).fit(X, y)

# Predict Income using C5.0 model
C5Pred = c5.predict(X)
C5Pred

array(['<=50K', '<=50K', '<=50K', ..., '<=50K', '<=50K', '<=50K'],
      dtype=object)

## 24. Evaluate Model 1 using the test data set. Construct a contingency table to compare the actual and predicted values of Income.

In [10]:
y_t = adult_test[['Income']]
X_t = adult_test[['Marital status', 'Cap_Gains_Losses']]

# Convert categorical variable into a dummy variable form
mar_dummy_t = pd.get_dummies(X_t['Marital status'])

#Add dummy variables back into X variables
X_t = pd.concat((X_t[['Cap_Gains_Losses']], mar_dummy_t), axis = 1) #Column-wise Concatation (axis=1)

# CART alogrithn (criterion = gini)
c5_t = DecisionTreeClassifier(criterion='entropy', max_leaf_nodes=5).fit(X_t, y_t)


# Predict Income using Cart
C5Pred_t = c5_t.predict(X_t)

# Confusion Matrix / Contigency Table 
cm = confusion_matrix(y_t, C5Pred_t)


TN = cm[0][0]
FP = cm[0][1]
FN = cm[1][0]
TP = cm[1][1]
print('TN: ', TN, '\nFP: ', FP, '\nFN: ', FN, '\nTP: ', TP)
cm

TN:  4627 
FP:  47 
FN:  1141 
TP:  340


array([[4627,   47],
       [1141,  340]])

## 25. For Model1, recapitulate Table7.4 from the text, calculating all of the model evaluation measures shown in the table. Call this table the Model Evaluation Table. Leave space for Model 2.

In [18]:
### Evaluation Measures ###
TAN = TN + FP
TAP = FN + TP
TPN = TN + FN
TPP =  FP + TP
GT = TN + FP + FN + TP

# Accuracy
Acc = (TN + TP) / (GT)
# Error Rate
Error = 1 - Acc
# Sensitivity / Recall
Sens = TP / TAP
#Specificity
Spec = TN/ TAN
# Precision
Prec = TP/TPP
# F1
F1 = 2*[(Prec*Sens)/(Prec+Sens)]
# F2
F2 = 5*[(Prec*Sens)/(4* Prec +Sens)]

print('Accuracy: ', Acc, '\nError Rate: ', Error, '\nSensitivity/Recall: ', Sens, '\nSpecificity: ', Spec, '\nPrecision: ', Prec)



Accuracy:  0.8069861900893582 
Error Rate:  0.19301380991064176 
Sensitivity/Recall:  0.22957461174881835 
Specificity:  0.9899443731279418 
Precision:  0.8785529715762274


## 26. Clearly and completely interpret each of the Model 1 evaluation measures from the Model Evaluation Table.

Accuracy: Accuracy represents an overall measure of the proportion of correct classifications made by the model. Model 1 acquired an 80.6% accuracy of correct classifications.

Error: Error rate measures the proportion of incorrect classifications. Model 1 had an error rate of 19.3%.

Sensitivity/Recall: Sensitivity measyres the ability of the model to classify positivitely. The specificity of the model is 22.29%. This is a poor percentage since only 22% of the actual positive records were classified as positive by the model

Specificity: Specificity measures what proportion of all negative records are captured in the model. The specificity of the model 98.89%. This is a great percentage since the model managed to correctyl classify of the actual negative records as negative.

Precision: The precision of the model is 87.8%. In other words, 87.8% of of records responded positively to the model.